### PR TITLE
improve docs

### PR DESCRIPTION
**Pull request by LiliC**
Opening this PR just to get some initial feedback and also to make the docs more of a collaborative effort. ;)

Will add comments where I need help/feedback on things.

(note to @krnowak will of course redo commits before merging.)

### CODE REVIEW

File: examples/crd.yml [hunk 1]
```diff
@@ -1,22 +0,0 @@
-apiVersion: apiextensions.k8s.io/v1beta1
```
Comment (LiliC) [inline examples/crd.yml]: Removed this file, as I did not see a real use for it. If you want to keep it, maybe would be best to go in the README?

File: README.md [hunk 2]
```diff
@@ -3,38 +3,38 @@
 
 # habitat-operator
 
-## Prerequisites
+This project is currently in `pre-alpha`.
```
Comment (LiliC) [inline README.md]: Added the project status, think we are in `pre-alpha`, wdyt?

File: README.md [hunk 3]
```diff
@@ -3,38 +3,38 @@
...
+
+## Overview
+
+Habitat operator is a Kubernetes controller designed to solve running and auto-managing Habitat Services on Kubernetes. It does this by making use of the [`Custom Resource Definition`][crd]s.
```
Comment (LiliC) [inline README.md]: Need more help/suggestions here, how to best describe what the operator does.

File: README.md [hunk 4]
```diff
@@ -3,38 +3,38 @@
...
-progress on upstreaming the feature.
+Note: Issues on this repo, are a way for developers to track ongoing work. The project issue tracker is located [here](https://github.com/kinvolk/habitat-operator/projects/1).
+But feel free to open an issue if you find something wrong or to request a feature.
 
-This means that images need to be created using the aforementioned fork of the
-`hab` client, which can be accomplished using [this
-script](https://gist.github.com/krnowak/3c854e94245e2f33a8366e629bfb09c8).
+## Prerequisites
+
+- Kubernetes cluster of version `>= 1.7.0`.
```
Comment (LiliC) [inline README.md]: Any other prerequisites we can think of?

File: examples/README.md [hunk 5]
```diff
@@ -0,0 +1,5 @@
+# Examples
+
```
Comment (LiliC) [inline examples/README.md]: More could be said here, but I do not know what else, any suggestions?

File: examples/namespaced/service_group.yml [hunk 6]
```diff
@@ -9,10 +9,8 @@ metadata:
   name: example-standalone-service-group
   namespace: example-namespace
 spec:
-  # the core/nginx habitat service packaged as a Docker image
```
Comment (LiliC) [inline examples/namespaced/service_group.yml]: Removed the comments, as they have nothing to do with the `namespaced` example itself.

File: README.md (hunk 2 above)
Comment (asymmetric) [reply README.md]: * I think it's somehow arbitrary, but
* Pre-alpha? That's harsh. Aren't we at least alpha? :cry:

Comment (LiliC) [reply README.md]: @asymmetric Not sure why that would be harsh, we have not cut a release yet, so we are in pre-alpha. When we have tested a Habitat service running for longer period of time we can cut an alpha release.

Comment (LiliC) [reply README.md]: Also I think it's important to convey that the operator is not production ready by displaying the project status.

Comment (krnowak) [reply README.md]: I'd say "This project is currently unstable - breaking changes may still land in future."

File: README.md [hunk 7]
```diff
@@ -3,38 +3,38 @@
...
+- At the moment, the Habitat operator requires a forked version of the Habitat supervisor that adds support for the `--peer-watch-file` flag.
+See [this issue](https://github.com/habitat-sh/habitat/issues/2735) to track progress on upstreaming the feature. This means that images need to be created using the aforementioned fork of the `hab` client, which can be accomplished using [this script](https://gist.github.com/krnowak/3c854e94245e2f33a8366e629bfb09c8).
```
Comment (krnowak) [inline README.md]: I'd rather point them to krnowak/hab stuff in habitat depot - https://bldr.habitat.sh/#/pkgs/krnowak/hab, the script itself just builds the HEAD of the habitat repo and uploads it to the depot.

File: README.md [hunk 8]
```diff
@@ -44,29 +44,37 @@ This will produce a `kinvolk/habitat-operator` image, which can then be deployed
...
+    kubectl create -f examples/standalone/service_group.yml
 
-To run end-to-end tests you need to have `minikube` up and running. After that just run:
- 
-    make TESTIMAGE=YOUR_OPERATOR_IMAGE e2e
-
-Clean up after the tests with:
-
-    make clean-test
+This will create a 1-pod deployment of an `nginx` Habitat service.
```
Comment (krnowak) [inline README.md]: a single pod deployment?

File: README.md [hunk 9]
```diff
@@ -44,29 +44,37 @@ This will produce a `kinvolk/habitat-operator` image, which can then be deployed
...
 
+### Testing
+
+To run unit tests locally, run:
+
+    make test
+
+To run end-to-end tests locally you need to have `minikube` up and running. After that just run:
+ 
+    make TESTIMAGE=YOUR_OPERATOR_IMAGE e2e
```
Comment (krnowak) [inline README.md]: What do we pass here in our tests? Maybe you could recommend that?

File: docs/api.md [hunk 10]
```diff
@@ -0,0 +1,45 @@
+# API
+
+The following is a description of the Habitat operator API. To see manifest example files, have a look at the [examples directory](https://github.com/kinvolk/habitat-operator/tree/master/examples).
+
+## ServiceGroup
+
```
Comment (krnowak) [inline docs/api.md]: Is this spacing required here?

File: examples/README.md (hunk 5 above)
Comment (krnowak) [reply examples/README.md]: Maybe mention that each example has their own readme.

File: examples/config/README.md [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+This example demostrates how initial configuration works with the Habitat operator. With the manifest file we deploy a `"Hello world."` Node.js Habitat service.
+We recommend that in the `default.toml` you do not write any secret configuration as this will be uploaded as a docker image. Instead use the initial configuration `user.toml` file.
+
+## Workflow
+
+After the Habitat operator is up and running, execute the following command from the root of this repository:
+
+`kubectl create -f /examples/nodejs`
+
+This will create a [Kubernetes Secret](https://kubernetes.io/docs/concepts/configuration/secret/) which will be mounted in correct place for Habitat to use it and override the configurations on initial launch.
```
Comment (krnowak) [inline examples/config/README.md]: which will be put in the correct place for…

I would avoid delving in such details like mounting - it is slightly more complicated than that. ;)

File: examples/config/README.md [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+It will also create a simple Node.js application that will display a msg. When running on minikube, it can be accessed under port `30001` of the minikube VM. `minikube ip` can be used to retrieve the IP.
+Initially our app is configured to display the msg `"Hello world."`. Because we override this with the Secret we just created, our app will instead display `Hello from our Habitat-Operator!`.
+
+## Deletion
+
+The Habitat operator does not delete the Secret on ServiceGroup deletion, as it is not managed by the Habitat operator.
```
Comment (krnowak) [inline examples/config/README.md]: Maybe just add a command that would do it then?

File: examples/encrypted/README.md [hunk 13]
```diff
@@ -1,5 +1,7 @@
 # Encrypted ServiceGroup example
 
+By default supervisors will communicate with no encryption. This example demostrates how to secure that connection.
```
Comment (krnowak) [inline examples/encrypted/README.md]: how to secure the communication

File: examples/encrypted/README.md [hunk 14]
```diff
@@ -19,11 +21,16 @@ For example, for a key named `foobar`, the key file might be something like
...
+After the Habitat operator is up and running, execute the following command from the root of this repository:
+
+```
+kubectl create -f examples/encrypted/service_group.yml
+```
 
 ## Deletion
 
-The operator does not delete the Secret on ServiceGroup deletion. This is
+The Habitat operator does not delete the Secret on ServiceGroup deletion. This is
```
Comment (krnowak) [inline examples/encrypted/README.md]: Maybe add a command to remove it?

File: examples/leader/README.md [hunk 15]
```diff
@@ -0,0 +1,15 @@
+# Leader topology example
+
+A topology describes the intended relationship between members within a Habitat service group. The `leader` topology employs leader election to define a leader.
```
Comment (krnowak) [inline examples/leader/README.md]: employs an election to choose a leader

File: examples/leader/README.md [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+
+## Workflow
+
+Simply run:
+
+  `kubectl create -f examples/leader/service_group.yml`.
+
+This will deploy 3 instances of consul Habitat service.
+
+Note: Whenever creating a `leader` topology ServiceGroup you have to specify `count: 3`, this is so election can take place.
```
Comment (krnowak) [inline examples/leader/README.md]: 3 or more and it would be best if the number is odd

File: README.md (hunk 2 above)
Comment (LiliC) [reply README.md]: A lot of projects have a clear pre-defined wording for the status of the project. `pre-alpha` means unstable, that is why I choose that. 

It's up to @asymmetric to choose the wording here, democracy wins. ;)

File: examples/namespaced/README.md [hunk 17]
```diff
@@ -1,8 +1,11 @@
 # Namespaced ServiceGroup example
 
+This demostrates how to deploy service in an [Kubernetes namespace](https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/).
```
Comment (krnowak) [inline examples/namespaced/README.md]: demonstrates
in a Kubernetes namespace

File: examples/standalone/README.md [hunk 18]
```diff
@@ -0,0 +1,11 @@
+# Standalone topology example
+
+A topology describes the intended relationship between members within a Habitat service group.
```
Comment (krnowak) [inline examples/standalone/README.md]: Maybe point to the habitat documentation about topologies?

File: examples/leader/README.md (hunk 15 above)
Comment (krnowak) [reply examples/leader/README.md]: Maybe add a link to the habitat documentation about topologies, leader election and whatnot.

File: README.md (hunk 7 above)
Comment (LiliC) [reply README.md]: @krnowak Sure, can you provide some instructions what to do with that depot?

File: README.md (hunk 9 above)
Comment (LiliC) [reply README.md]: We pass in the image of the operator? :)

File: examples/leader/README.md (hunk 16 above)
Comment (LiliC) [reply examples/leader/README.md]: Really, did not know that last part, cool.

File: README.md (hunk 9 above)
Comment (krnowak) [reply README.md]: Yeah, I was just wondering where does this image come from. Is it something we uploaded to docker registry (if yes, maybe mention the name of it, so others can use it) or something we build (how?)

File: README.md (hunk 7 above)
Comment (krnowak) [reply README.md]: I think you can basically borrow stuff you pinned on slack. This `hab pkg install krnowak/hab`, `binlink` and so on.

File: README.md (hunk 9 above)
Comment (LiliC) [reply README.md]: Yes we above mention how to build the image. And give really detailed instructions for it. https://github.com/kinvolk/habitat-operator/blob/3965f63ffffca5378fea815123c43fe28e841986/README.md#running-inside-a-kubernetes-cluster

Maybe we can say, `to build the image see` and link to that? But it is in the same README, so I am not sure its needed?

File: examples/leader/README.md (hunk 16 above)
Comment (krnowak) [reply examples/leader/README.md]: Yeah, it is to avoid a scenario where half of the votes went to one supervisor and another half to another one, so there is no clear winner. Unlikely, but still.

File: README.md (hunk 9 above)
Comment (krnowak) [reply README.md]: I just read the README.md. No need to do anything here. Sorry for the noise.

File: README.md (hunk 7 above)
Comment (LiliC) [reply README.md]: I updated the link to the script, wdyt?

File: README.md (hunk 4 above)
Comment (asymmetric) [reply README.md]: I'd just reword this as "Kubernetes >= 1.7.0", the rest is redundant IMO.

File: README.md (hunk 2 above)
Comment (asymmetric) [reply README.md]: I don't know if the pre-alpha thing is that standardized, that's why I think it's somehow arbitrary.

But if we can agree that it is in pre-alpha, according to some criteria, then by all means :)

Otherwise I'd rather go with @krnowak's more "fluid" option.

File: README.md [hunk 19]
```diff
@@ -20,7 +20,7 @@ But feel free to open an issue if you find something wrong or to request a featu
 
 - Kubernetes cluster of version `>= 1.7.0`.
 - At the moment, the Habitat operator requires a forked version of the Habitat supervisor that adds support for the `--peer-watch-file` flag.
-See [this issue](https://github.com/habitat-sh/habitat/issues/2735) to track progress on upstreaming the feature. This means that images need to be created using the aforementioned fork of the `hab` client, which can be accomplished using [this script](https://gist.github.com/krnowak/3c854e94245e2f33a8366e629bfb09c8).
+See [this issue](https://github.com/habitat-sh/habitat/issues/2735) to track progress on upstreaming the feature. This means that images need to be created using the aforementioned fork of the `hab` client, which can be accomplished following [this script](https://gist.github.com/LiliC/c028fc4687f466e3e3bd5981d2529173).
```
[the remark this answers is not in the document]
Comment (LiliC) [reply README.md]: No, not that I am aware of. Can add it to the `README`, just felt it makes it less readable for users. WDYT?

File: README.md (hunk 2 above)
Comment (LiliC) [reply README.md]: Sure will use @krnowak version. FYI: https://en.wikipedia.org/wiki/Software_release_life_cycle#Pre-alpha :)

File: README.md [hunk 20]
```diff
@@ -3,38 +3,38 @@
 
 # habitat-operator
 
-## Prerequisites
+This project is currently in `pre-alpha`.
+
+## Overview
 
-The Habitat Operator makes use of [`Custom Resource Definition`][crd]s, and requires a Kubernetes cluster of version `>= 1.7.0`.
+Habitat operator is a Kubernetes controller designed to solve running and auto-managing Habitat Services on Kubernetes. It does this by making use of the [`Custom Resource Definition`][crd]s.
```
Comment (asymmetric) [inline README.md]: Why remove the article at the beginning? It reads weird to me like this.

Comment (asymmetric) [reply README.md]: I would say "by making use of Custom...", i.e. without the "the".

File: README.md [hunk 21]
```diff
@@ -3,38 +3,38 @@
...
 
-At the moment, the Operator requires a forked version of the Habitat supervisor
-that adds support for the `--peer-watch-file` flag.
-See [this issue](https://github.com/habitat-sh/habitat/issues/2735) to track
-progress on upstreaming the feature.
+Habitat is an application automation framework that allows you to build applications that have automation built-in. To learn more about Habitat, please visit the [Habitat website](https://www.habitat.sh/).
```
Comment (asymmetric) [inline README.md]: Don't think we need to explain what Habitat is here, a link to their page should be enough.

File: README.md [hunk 22]
```diff
@@ -3,38 +3,38 @@
...
 
-This means that images need to be created using the aforementioned fork of the
-`hab` client, which can be accomplished using [this
-script](https://gist.github.com/krnowak/3c854e94245e2f33a8366e629bfb09c8).
+For more detailed description of the Habitat operator API have a look at the [API documentation](https://github.com/kinvolk/habitat-operator/blob/1c2dbfa16c4163d1c185e76f1821a9a542acb066/docs/api.md).
```
Comment (asymmetric) [inline README.md]: For a more.

Comment (asymmetric) [reply README.md]: Couldn't the link just point to the master tree in this case? Otherwise it will point to outdated versions in the future.

File: README.md [hunk 23]
```diff
@@ -3,38 +3,38 @@
...
+
+Note: Issues on this repo, are a way for developers to track ongoing work. The project issue tracker is located [here](https://github.com/kinvolk/habitat-operator/projects/1).
```
Comment (asymmetric) [inline README.md]: I think this is unnecessary, as that's how most projects work IMO.

File: README.md [hunk 24]
```diff
@@ -44,29 +44,37 @@ This will produce a `kinvolk/habitat-operator` image, which can then be deployed
...
+    kubectl create -f examples/standalone/service_group.yml
 
-To run end-to-end tests you need to have `minikube` up and running. After that just run:
- 
-    make TESTIMAGE=YOUR_OPERATOR_IMAGE e2e
-
-Clean up after the tests with:
-
-    make clean-test
+This will create a single pod deployment of an `nginx` Habitat service.
```
Comment (asymmetric) [inline README.md]: I'd say "single-pod"

File: docs/api.md [hunk 25]
```diff
@@ -0,0 +1,37 @@
...
+| Field | Description | Scheme | Required |
```
Comment (asymmetric) [inline docs/api.md]: I'm a little bit concerned this will become hard to maintain. Have you used a tool to build it?

File: examples/config/README.md [hunk 26]
```diff
@@ -0,0 +1,22 @@
+# Initial configuration
+
+This example demostrates how initial configuration works with the Habitat operator. With the manifest file we deploy a `"Hello world."` Node.js Habitat service.
+We recommend that in the `default.toml` you do not write any secret configuration as this will be uploaded as a docker image. Instead use the initial configuration `user.toml` file.
```
Comment (asymmetric) [inline examples/config/README.md]: I'd remove the personal pronouns: "**NOTE**: Adding secret configuration to the `default.toml` is discouraged, as it will be uploaded..."

File: examples/crd.yml (hunk 1 above)
Comment (asymmetric) [reply examples/crd.yml]: This is useful when testing manually, you can remove and create the crd with `kubectl`.

I'd leave it here, if it doesn't bother you ;)

File: examples/encrypted/README.md [hunk 27]
```diff
@@ -1,5 +1,7 @@
 # Encrypted ServiceGroup example
 
+By default supervisors will communicate with no encryption. This example demostrates how to secure the communication.
```
Comment (asymmetric) [inline examples/encrypted/README.md]: :+1:

File: examples/namespaced/service_group.yml (hunk 6 above)
Comment (asymmetric) [reply examples/namespaced/service_group.yml]: In which example would you leave them?

File: README.md (hunk 2 above)
Comment (asymmetric) [reply README.md]: Yeah, that's the enterprise-y thing I mentioned in my comment. Not sure who adheres to that.

File: README.md (hunk 20 above)
Comment (LiliC) [reply README.md]: I cannot do my `the`'s :).

File: README.md (hunk 22 above)
Comment (LiliC) [reply README.md]: Good catch, forgot to change it when I copy pasted!

File: README.md (hunk 23 above)
Comment (LiliC) [reply README.md]: Habitat also has this note in their project, but can remove it from here?

File: README.md (hunk 24 above)
Comment (LiliC) [reply README.md]: @krnowak are you okay with `single-pod` as you suggested `single pod`. 😄

File: docs/api.md (hunk 25 above)
Comment (LiliC) [reply docs/api.md]: I am not sure we will be adding any new fields anytime soon, or at least we do not have any planned?

File: examples/crd.yml (hunk 1 above)
Comment (LiliC) [reply examples/crd.yml]: Can leave sure. Just not sure I get what you mean by testing manually, I am interested as I never used this file before :)

File: examples/namespaced/service_group.yml (hunk 6 above)
Comment (LiliC) [reply examples/namespaced/service_group.yml]: We have them in other examples, this just demonstrates deploying a SG to a namespace so it does not matter what the image contains. But can undo the removal.

File: README.md (hunk 24 above)
Comment (LiliC) [reply README.md]: from offline: he is okay with it.

File: examples/namespaced/service_group.yml (hunk 6 above)
Comment (LiliC) [reply examples/namespaced/service_group.yml]: I noticed you later edited your comment. Would prefer if you just write another comment instead of editing the original, as this is much easier to notice.

Like I mentioned I can undo the removal. 👍

**Comment by LiliC** [comment]
@asymmetric Adjusted based on your suggestions.

File: README.md (hunk 23 above)
Comment (asymmetric) [reply README.md]: Yeah, I think it's confusing, because in the end we can deal with both types of issues I think. Also, they aren't so different in a way. It's just stuff that has to be done according to someone.

Sometimes I might open an issue, and you might say it's invalid for some reason and we would close it, etc. etc.

File: docs/api.md (hunk 25 above)
Comment (asymmetric) [reply docs/api.md]: Yeah, maybe. Let's see. If it becomes hard to maintain I will propose to remove it :)

File: examples/crd.yml (hunk 1 above)
Comment (asymmetric) [reply examples/crd.yml]: For example, if you remove the whole CRD resource and you can re-create it with this file. It has been useful for me in the past, and also served as a sort of easier-to-read documentation. Now we have that in the README, true.

Don't know, I'd leave it :)

File: docs/api.md (hunk 25 above)
Comment (LiliC) [reply docs/api.md]: Sgtm.

File: README.md (hunk 23 above)
Comment (LiliC) [reply README.md]: Will remove. 👍

**Comment by LiliC** [comment]
Rebased on latest master and merged into one commit. Will write sensible commits once @krnowak has a look again in case he has anything else to suggest.

File: README.md [hunk 28]
```diff
@@ -3,38 +3,35 @@
...
 
-This means that images need to be created using the aforementioned fork of the
-`hab` client, which can be accomplished using [this
-script](https://gist.github.com/krnowak/3c854e94245e2f33a8366e629bfb09c8).
+For a more detailed description of the Habitat operator API have a look at the [API documentation](https://github.com/kinvolk/habitat-operator/blob/master/docs/api.md).
+
+## Prerequisites
+
+- Kubernetes `>= 1.7.0`.
+- At the moment, the Habitat operator requires a forked version of the Habitat supervisor that adds support for the `--peer-watch-file` flag.
```
Comment (asymmetric) [inline README.md]: @blixtra was saying that this should be moved to the top, so that it's more prominent. WDYT?

Comment (LiliC) [reply README.md]: By this you or rather @blixtra means just the part about the operator requiring the forked version or the whole section?

File: examples/README.md [hunk 29]
```diff
@@ -0,0 +1,7 @@
...
+To see all avaliable fields have a look at the [API docs](https://github.com/kinvolk/habitat-operator/blob/master/docs/api.md).
+
+To run any of these examples, the Habitat operator must be up and running. To achieve that just follow this [usage documention](https://github.com/kinvolk/habitat-operator#usage).
+
+Note each example has their own documentation.
```
Comment (krnowak) [inline examples/README.md]: Note that each

Comment (krnowak) [reply examples/README.md]: Or "Note: Each"

Comment (LiliC) [reply examples/README.md]: "Nit" addressed. ;)

File: README.md (hunk 28 above)
Comment (asymmetric) [reply README.md]: Only the part about the fork.

File: README.md [hunk 30]
```diff
@@ -3,38 +3,35 @@
 
 # habitat-operator
 
-## Prerequisites
+This project is currently unstable - breaking changes may still land in future.
```
Comment (asymmetric) [inline README.md]: in *the* future ;)

**Comment by LiliC** [comment]
@asymmetric @krnowak Fixed the docs as agreed, can you have another look.

**Comment by krnowak** [comment]
LFAD.
